src/fsharp/FSharp.Core/array.fs

```diff
@@ -483,19 +483,45 @@ namespace Microsoft.FSharp.Collections
                             count <- count + 1
             Array.subUnchecked 0 count res
 
-
         [<CompiledName("Filter")>]
-        let filter f (array: _[]) = 
-            checkNonNull "array" array
-            let res = Array.zeroCreateUnchecked array.Length 
-            let mutable count = 0
-            for x in array do                 
-                if f x then 
-                    res.[count] <- x
-                    count <- count + 1
-            Array.subUnchecked 0 count res
-
+        let filter f (array: _[]) =         
+            checkNonNull "array" array                    
+            let mutable i = 0    
+            while i < array.Length && not (f array.[i]) do
```

Comment:
This is a really neat optimization, it means that filters that find nothing allocate nothing.  And filters that don't hit early in the iteration allocate fewer elements.  And it doesn't even do more work to do that.

I like this a lot.

Kevin